keps/sig-scheduling/3902-decoupled-taint-manager/README.md

```diff
@@ -0,0 +1,300 @@
+
+In this KEP, we propose to 
+
+* Decouple `TaintManager` that performs taint-based pod eviction from `NodeLifecycleController` and make them two separate controllers: `NodeLifecycleManager` to add taints to unhealthy nodes and `TaintManager` to perform pod deletion on nodes tainted with NoExecute effect.
+
+* Allow users to opt-out from the taint-based eviction while still retaining the overall functionality provided by `NodeLifecycleController.`
+
+## Motivation
+* `NodeLifecycleController` combines two independent functions: adding a pre-defined set of `NoExecute` taints to unhealthy nodes and performing pod eviction on arbitrary `NoExecute` taints. While  `NodeLifecycleController`'s responsibility is primarily on Node unhealthy-zone, `TaintManager` is concerned with actioning the taints added by `NodeLifecycleController` by performing Pods eviction. De-coupling `NoExecuteTaintManager` from `NodeLifecycleController` in `kube-controller-manager` would allow to implement a single eviction controller to take care of general Pod eviction and deletion for different cases based on workload needs, including `NoExecute` taint-based pod eviction other than node `Unhealthy` and `Unreachable`.  Splitting `NodeLifecycleController` into two separate controllers will support more flexible extension and improvement of the existing node taint-based pod eviction. 
+* There are increasing needs to customize and extend the default node taint-based pod evictions in practice.
+  * Whether or not evict a stateful pod with local storage on `NoExecute` can depend on the actual taint conditions, workload status and other conditions. The decision may require information from the external systems and controllers, which have more knowledge and context about the workloads.  A custom `taint-manager`  will be able to control how and when a pod is evicted or deleted, including making calls to external systems and take actions on  `NoExecute` taints with other conditions such as unhealthy network. 
+  * Starting kubernetes 1.27, with the removal of the flag `enable-taint-manager`, replacing the default `NoExecute` taint-manager with a custom one would be more challenging. `NoExecute` taints can disable the no-execute taint manager, but such disabling can only be achieved through a mutating webhook on all pods, and it can interact negatively with the upstream `DefaultTolerations` webhook.
+  * The requirement of custom taint-based eviction is not unique. It exists already in Kubernetes, where we call it `FullyDistributionMode`. When all nodes are unhealthy, `NodeLifecycleController` chooses not to apply `NoExecute` taints properly. In a real-world case, the integrator may choose to define the disruption rate/budget, etc.
+
+### Goals
+
+* Move taint-based eviction implementation out of `NodeLifecycleController` into a separate and independent taint-manager to support flexible extension and customization of node taint-based eviction.
```

Comment:
```suggestion
* Move taint-based eviction implementation out of `NodeLifecycleController` into a separate and independent taint-manager to enhance separation of concerns and maintainability.
```

Reply:
The current wording can give the false impression that anyone can implement an extension, which is not true. Only users who have the ability to replace or configure control plane components can.

Reply:
Good point, we'll clarify that a only the cluster administrators can do it.

Reply:
It will monitor these information changes. If a node has not updated its status for a long time, then we can assume that an exception has occurred in this node, and some fault tolerance processing is needed to safely remove the pod on this node. These Pods are rebuilt on other nodes.